validator/src/machine/keygen/confirmed.ts

```diff
@@ -25,56 +31,105 @@ export const handleKeyGenConfirmed = async (
 	}
 	const groupId = event.gid;
 
-	// Check if this is our own confirmation
-	const ourParticipantId = keyGenClient.participantId(groupId);
-	if (event.identifier !== ourParticipantId) {
-		// Not our confirmation, just track
-		logger?.(`Group ${event.gid} confirmation from ${event.identifier} (not ours)`);
+	// Track this confirmation
+	const confirmedParticipants = [...machineStates.rollover.confirmedParticipants, event.identifier];
+	const participants = signingClient.participants(groupId);
+	const allConfirmed = confirmedParticipants.length === participants.length;
```

Comment:
An array does not prevent duplicated identifiers. In the spec we use a set (which I don't think is necessary for the validator implementation). I would check that for each participant we have a confirmation.

Reply:
Updated to check that every participant has an item in `sharesFrom`.